viewing_party/party.py

```diff
@@ -0,0 +1,140 @@
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+    
+    return user_data
+
+
+def get_watched_avg_rating(user_data):
+    avg_rating = 0.0
+    total_rating = 0.0
+    
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"] 
+    
+    if len(user_data["watched"]) > 0:
+        avg_rating = total_rating/len(user_data["watched"])
+    return avg_rating
+
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+        return None
+    
+    genre_dict = {}
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in genre_dict:
+            genre_dict[genre] += 1
+        else:
+            genre_dict[genre] = 1
+
+    most_freq = genre
+    for genre, freq in genre_dict.items():
+        if freq > genre_dict[most_freq]:
+            most_freq = genre
+
+    return most_freq
+
+
+def get_unique_watched(user_data):
+
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie)
+    
+    friends_watched = []
+    for watched_dict in user_data["friends"]:
+        for title in watched_dict["watched"]:
+            friends_watched.append(title)
+
+    user_unique_watched = []
+    for movie in user_watched:
+        if movie not in friends_watched:
+            user_unique_watched.append(movie)
+
+    return user_unique_watched
+
+
+def get_friends_unique_watched(user_data):
+    
+    friends_watched = []
+    for watched_dict in user_data["friends"]:
+        for title in watched_dict["watched"]:
+            friends_watched.append(title)
+    
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie)
+    
+
+    friends_unique_watched = []
+    for movie in friends_watched:
+        if movie not in friends_unique_watched:
+            movie_in_user_watched = False
+            for title in user_watched: 
+                if movie["title"] == title["title"]:
+                    movie_in_user_watched = True
+
+            if not movie_in_user_watched:
+                friends_unique_watched.append(movie)
+
+    return friends_unique_watched
+
+
+def get_available_recs(user_data):
+    watchlist = get_friends_unique_watched(user_data)
+
+    movie_recs = []
+    for movie in watchlist:
+        if movie["host"] in user_data["subscriptions"]:
+            movie_recs.append(movie)
+    
+    return movie_recs
+
+
+def get_new_rec_by_genre(user_data):
+    fav_genre = get_most_watched_genre(user_data)
+    available_recs = get_friends_unique_watched(user_data)
+
+    movie_recs = []
+    for movie in available_recs:
+        if fav_genre == movie["genre"]:
+            movie_recs.append(movie)
+    return movie_recs
+
+
+def get_rec_from_favorites(user_data):
+    unique_watched = get_unique_watched(user_data)
+    fav_recs = []
+
+    for movie in user_data["favorites"]:
+        if movie in unique_watched:
```

Comment:
We could either extract a set of titles from the unique watched result, or build it from the favorites collection itself. In either case, we would iterate over the other collection, and check whether the item titles in that other collection are in the title set. If we wanted to make the title set from the unique watched result, we could even use comprehension syntax to do so!
```python
title_set = {movie["title"] for movie in unique_watched}
```

Incidentally, the reason this could be done starting from either collection is because here we are basically looking for a union between the user's favorite movies, and movies only they have watched. A union is a commutative operation (the order of the inputs doesn't matter). For contrast, difference (or subtraction for regular numbers) is _not_ commutative, and so the order matters. e.g. 3 - 1 doesn't equal 1 - 3.